Mathlib/Data/Matroid/Rank/Finite.lean

```diff
@@ -0,0 +1,190 @@
+lemma rkFin_iff_exists_basis' : M.RkFin X ↔ ∃ I, M.Basis' I X ∧ I.Finite :=
+  ⟨RkFin.exists_finite_basis', fun ⟨_, hIX, hI⟩ ↦ hIX.rkFin_of_finite hI⟩
+
+lemma RkFin.subset (h : M.RkFin X) (hXY : Y ⊆ X) : M.RkFin Y := by
+  obtain ⟨I, hI⟩ := M.exists_basis' Y
+  obtain ⟨J, hJ, hIJ⟩ := hI.indep.subset_basis'_of_subset (hI.subset.trans hXY)
+  exact hI.rkFin_of_finite <| (hJ.finite_of_rkFin h).subset hIJ
+
+@[simp]
+lemma rkFin_inter_ground_iff : M.RkFin (X ∩ M.E) ↔ M.RkFin X :=
+  let ⟨_I, hI⟩ := M.exists_basis' X
+  ⟨fun h ↦ hI.rkFin_of_finite (hI.basis_inter_ground.finite_of_rkFin h),
+    fun h ↦ h.subset inter_subset_left⟩
+
+lemma RkFin.to_inter_ground (h : M.RkFin X) : M.RkFin (X ∩ M.E) :=
+  rkFin_inter_ground_iff.2 h
+
+lemma rkFin_iff (hX : X ⊆ M.E := by aesop_mat) : M.RkFin X ↔ ∃ I, M.Basis I X ∧ I.Finite := by
+  simp_rw [rkFin_iff_exists_basis', M.basis'_iff_basis hX]
+
+lemma Indep.rkFin_iff_finite (hI : M.Indep I) : M.RkFin I ↔ I.Finite :=
+  hI.basis_self.finite_iff_rkFin.symm
+
+lemma Indep.finite_of_rkFin (hI : M.Indep I) (hfin : M.RkFin I) : I.Finite :=
+  hI.basis_self.finite_of_rkFin hfin
+
+lemma rkFin_of_finite (M : Matroid α) (hX : X.Finite) : M.RkFin X :=
+  let ⟨_, hI⟩ := M.exists_basis' X
+  hI.rkFin_of_finite (hX.subset hI.subset)
+
+lemma Indep.subset_finite_basis'_of_subset_of_rkFin (hI : M.Indep I) (hIX : I ⊆ X)
+    (hX : M.RkFin X) : ∃ J, M.Basis' J X ∧ I ⊆ J ∧ J.Finite :=
+  (hI.subset_basis'_of_subset hIX).imp fun _ hJ => ⟨hJ.1, hJ.2, hJ.1.finite_of_rkFin hX⟩
+
+lemma Indep.subset_finite_basis_of_subset_of_rkFin (hI : M.Indep I) (hIX : I ⊆ X)
+    (hX : M.RkFin X) (hXE : X ⊆ M.E := by aesop_mat) : ∃ J, M.Basis J X ∧ I ⊆ J ∧ J.Finite :=
+  (hI.subset_basis_of_subset hIX).imp fun _ hJ => ⟨hJ.1, hJ.2, hJ.1.finite_of_rkFin hX⟩
+
+lemma rkFin_singleton (M : Matroid α) (e : α) : M.RkFin {e} :=
+  rkFin_of_finite M (finite_singleton e)
+
+@[simp]
+lemma RkFin.empty (M : Matroid α) : M.RkFin ∅ :=
+  rkFin_of_finite M finite_empty
+
+lemma not_rkFin_superset (h : ¬M.RkFin X) (hXY : X ⊆ Y) : ¬M.RkFin Y :=
+  fun h' ↦ h (h'.subset hXY)
+
+lemma RkFin.finite_of_indep_subset (hX : M.RkFin X) (hI : M.Indep I) (hIX : I ⊆ X) : I.Finite :=
+  hI.finite_of_rkFin (hX.subset hIX)
+
+@[simp]
+lemma rkFin_ground_iff_rankFinite : M.RkFin M.E ↔ M.RankFinite := by
+  rw [RkFin, restrict_ground_eq_self]
+
+lemma rkFin_ground (M : Matroid α) [RankFinite M] : M.RkFin M.E := by
+  rwa [rkFin_ground_iff_rankFinite]
+
+lemma Indep.finite_of_subset_rkFin (hI : M.Indep I) (hIX : I ⊆ X) (hX : M.RkFin X) : I.Finite :=
+  hX.finite_of_indep_subset hI hIX
+
+lemma RkFin.closure (h : M.RkFin X) : M.RkFin (M.closure X) :=
+  let ⟨_, hI⟩ := M.exists_basis' X
+  hI.basis_closure_right.rkFin_of_finite <| hI.finite_of_rkFin h
+
+@[simp]
+lemma rkFin_closure_iff : M.RkFin (M.closure X) ↔ M.RkFin X := by
+  rw [← rkFin_inter_ground_iff (X := X)]
+  exact ⟨fun h ↦ h.subset <| M.inter_ground_subset_closure X, fun h ↦ by simpa using h.closure⟩
```

Comment:
```suggestion
  exact ⟨fun h ↦ h.subset <| M.inter_ground_subset_closure X, fun h ↦ h.closure⟩
```
looks like this should work?

Reply:
It doesn't, because of the rewrite in the first line.